File: event/programmers/2019-jobfair/3.js
function solution(sticker) {
  const dp = Array(sticker.length).fill(null)

  const queue = []
  let max = 0

  queue.push([0, sticker[0]]) // currIdx , sum
  queue.push([1, sticker[1]]) // currIdx , sum

  while (queue.length) {
    const [currIdx, sum] = queue.shift()

    if (dp[currIdx] < sum) {
      max = max > sum ? max : sum
      dp[currIdx] = sum

      if (currIdx + 2 < sticker.length) {
        queue.push([currIdx + 2, sum + sticker[currIdx + 2]])
      }
      if (currIdx + 3 < sticker.length) {
        queue.push([currIdx + 3, sum + sticker[currIdx + 3]])
      }
    }
  }
  return max
}
//////////////////////////

const { algoTest } = require('../../../utils')

const testFunc = solution
const params = [
  [[12]], //
  [[12, 1, 12]], //
  [[12, 12, 12, 12, 12]], //
  [[12, 80, 14, 22, 100]],
]
const expects = [
  12, //
  24, //
  36, //
  180,
]

algoTest(testFunc, params, expects)

File: algorithm/event/programmers/2019-jobfair/3.py
import collections


def solution(sticker):
    dp = [0] * len(sticker)
    queue = collections.deque()
    max = 0

    queue.append((0, sticker[0]))
    queue.append((1, sticker[1]))

    while queue:
        (sticket_idx, sum) = queue.popleft()
        if dp[sticket_idx] < sum:

            max = max if max > sum else sum
            dp[sticket_idx] = sum

            if (sticket_idx + 2 < len(sticker)):
                queue.append([sticket_idx + 2, sum + sticker[sticket_idx + 2]])

            if (sticket_idx + 3 < len(sticker)):
                queue.append([sticket_idx + 3, sum + sticker[sticket_idx + 3]])

    return max


print(solution([12, 12, 12, 12, 12]), 36)
print(solution([12, 80, 14, 22, 100]), 180)
